x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -732,40 +786,71 @@ private void innerRefresh(String tokenDocId, Authentication userAuth, ActionList
                         final Map<String, Object> source = response.getSource();
                         final Optional<ElasticsearchSecurityException> invalidSource = checkTokenDocForRefresh(source, userAuth);
-
                         if (invalidSource.isPresent()) {
                             onFailure.accept(invalidSource.get());
                         } else {
-                            final Map<String, Object> userTokenSource = (Map<String, Object>)
-                                ((Map<String, Object>) source.get("access_token")).get("user_token");
-                            final String authString = (String) userTokenSource.get("authentication");
-                            final Integer version = (Integer) userTokenSource.get("version");
-                            final Map<String, Object> metadata = (Map<String, Object>) userTokenSource.get("metadata");
-
-                            Version authVersion = Version.fromId(version);
-                            try (StreamInput in = StreamInput.wrap(Base64.getDecoder().decode(authString))) {
-                                in.setVersion(authVersion);
-                                Authentication authentication = new Authentication(in);
-                                UpdateRequestBuilder updateRequest =
-                                    client.prepareUpdate(SecurityIndexManager.SECURITY_INDEX_NAME, TYPE, tokenDocId)
-                                        .setDoc("refresh_token", Collections.singletonMap("refreshed", true))
-                                        .setRefreshPolicy(RefreshPolicy.WAIT_UNTIL);
-                                updateRequest.setIfSeqNo(response.getSeqNo());
-                                updateRequest.setIfPrimaryTerm(response.getPrimaryTerm());
-                                executeAsyncWithOrigin(client.threadPool().getThreadContext(), SECURITY_ORIGIN, updateRequest.request(),
-                                    ActionListener.<UpdateResponse>wrap(
-                                        updateResponse -> createUserToken(authentication, userAuth, listener, metadata, true),
-                                        e -> {
-                                            Throwable cause = ExceptionsHelper.unwrapCause(e);
-                                            if (cause instanceof VersionConflictEngineException ||
-                                                isShardNotAvailableException(e)) {
-                                                innerRefresh(tokenDocId, userAuth,
-                                                    listener, attemptCount);
-                                            } else {
-                                                onFailure.accept(e);
-                                            }
-                                        }),
-                                    client::update);
+                            if (eligibleForMultiRefresh(source)) {
```

Comment:
we probably want to capture an instant early on in the refresh code and pass it here. Otherwise during retries a valid request may not be allowed to be refreshed.

Reply:
An issue with this approach is that if we capture instants on when the refresh request arrives, we will have to deal with instants that are not simply in the `[refreshTime, refreshTime+4s]` timespan but with instants that are before `refreshTime` too.

Reply:
I have also been wondering about the 4s timeframe. I think we need to be conscious of things like a long GC or nodes without perfect time syncs. Maybe between `[refreshTime - 30s, refreshTime + 30s]`?

Reply:
Yes, agreed. I don't think that the 1min duration introduces any significant changes to the thread model described in https://github.com/elastic/elasticsearch/issues/36872#issuecomment-455250760

Although, the bigger problem here (I was about to add this as a general comment but fits here also) is "quite" the opposite. Tests show that with hitting many nodes concurrently, it's quite often the case where the 2nd,3rd refresh request get the token document source _before_ (even with refresh policy to IMMEDIATE on the update request) the first refresh has the chance to update it and thus `if (eligibleForMultiRefresh(source))` is false, and we proceed to create new tokens. 
I'll try to put the GetRequest for the document within the `innerRefresh()` again ( it's an extra Get request while we have the source already but.. ) in order to make sure that `eligibleForMultiRefresh` gets a fresher view of the document and hopefully the update is already reflected in there, even for nearly concurrent requests to refresh